core/utils/managed_allocator.hpp

```diff
@@ -22,79 +22,93 @@
 
 #pragma once
 
+#include <memory>
+#include <type_traits>
+
 #include "misc.hpp"
 
 namespace irs {
-template<typename Allocator, typename Manager>
-class ManagedAllocator : private Allocator {
+
+// It's possible to allow unsafe swap: swap without allocator propagation.
+// But beware of UB in case of non equal allocators.
+template<typename M, typename A, bool SafeSwap = true>
+class ManagedAllocator {
+  using Traits = std::allocator_traits<A>;
+
  public:
-  using difference_type =
-    typename std::allocator_traits<Allocator>::difference_type;
-  using propagate_on_container_move_assignment = typename std::allocator_traits<
-    Allocator>::propagate_on_container_move_assignment;
-  using propagate_on_container_copy_assignment = typename std::allocator_traits<
-    Allocator>::propagate_on_container_copy_assignment;
-  using propagate_on_container_swap = std::true_type;
-  // Note: If swap would be needed this code should do the trick.
-  // But beware of UB in case of non equal allocators.
-  // std::conditional_t<std::is_empty_v<Allocator>, std::true_type,
-  // typename std::allocator_traits<Allocator>::propagate_on_container_swap>;
-  using size_type = typename std::allocator_traits<Allocator>::size_type;
-  using value_type = typename std::allocator_traits<Allocator>::value_type;
+  using value_type = typename Traits::value_type;
+  using size_type = typename Traits::size_type;
+  using difference_type = typename Traits::difference_type;
+  using propagate_on_container_move_assignment =
+    typename Traits::propagate_on_container_move_assignment;
+  using propagate_on_container_copy_assignment =
+    typename Traits::propagate_on_container_copy_assignment;
```

Comment:
We could use inheritance from std::allocator_traits<A> instead of this.
But it doesn't look like a good idea for me.
At least because combining few empty classes, especially ebo with no_unique_address could work not as expected for some compilers